### PR TITLE
Tensor Parallelism Tests

### PR DESCRIPTION
**Pull request by eitanturok**
# What does this PR do?

Test tp-fsdp against fsdp and ddp on:
- [x] forward pass
- [x] gradients
- [x] loss
- [x] accuracy
- [ ] memory

I'll test memory in a follow up PR.

Other fixes:
* fix length of `SimpleDataset`

# What issue(s) does this change relate to?

# Before submitting
- [ ] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/main/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [x] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/main/CONTRIBUTING.md#running-tests))
- [x] Did you run the tests locally to make sure they pass?
- [x] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/main/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

**Review by snarayan21 — COMMENTED** [review]
seems some of the gpu tests are failing + linting btw

File: tests/trainer/test_tp.py [hunk 1]
```diff
@@ -1,27 +1,456 @@
...
+from composer.utils import FSDPConfig, ParallelismConfig, TPConfig, dist, reproducibility
 from tests.common import (
     RandomClassificationDataset,
+    SimpleComposerMLP,
     SimpleModel,
     world_size,
 )
 
 
+class RandomClassificationDatasetReplicated(Dataset):
```
Comment (mvpatel2000) [inline tests/trainer/test_tp.py:28]: should this be in datasets?

Comment (mvpatel2000) [reply tests/trainer/test_tp.py:28]: can this be a subclass for simplicity?

File: tests/trainer/test_tp.py [hunk 2]
```diff
@@ -1,27 +1,456 @@
...
+def _replace_state_dict_name(state_dict: dict[str, Any], old_name: str, new_name: str) -> dict[str, Any]:
+    keys = list(state_dict.keys())
+    for key in keys:
+        if old_name in key:
+            new_key = key.replace(old_name, new_name, 1)
+            state_dict[new_key] = state_dict.pop(key)
+    return state_dict
+
+
+def compare_modules(
```
Comment (mvpatel2000) [inline tests/trainer/test_tp.py:236]: isnt this implemented elsewhere in common?

File: tests/trainer/test_tp.py [hunk 3]
```diff
@@ -1,27 +1,456 @@
...
+    tp_fsdp_trainer.fit()
+    tp_fsdp_trainer.close()
+    tp_fsdp_stats = get_stats(tp_fsdp_trainer)
+
+    # Ensure the updated models weights are the same
+    # Drop tolerance due to precision issues across different parallelism strategies
+    compare_models(ddp_trainer, fsdp_trainer, tp_fsdp_trainer, atol=1e-5, rtol=1e-3)
+
+    # Compare loss between DDP, FSDP, TP-FSDP
+    np.testing.assert_allclose(
```
Comment (mvpatel2000) [inline tests/trainer/test_tp.py:407]: why np and not torch?

Comment (eitanturok) [reply tests/trainer/test_tp.py:407]: because the accuracy is saved as a np array by default in the in-memory logger

File: tests/trainer/test_tp.py (hunk 1 above)
Comment (eitanturok) [reply tests/trainer/test_tp.py:28]: i don't think subclassing would make things simpler as we'd have to overwrite every single method of `RandomClassificationDataset`

Comment (eitanturok) [reply tests/trainer/test_tp.py:28]: * moved to datasets
* I subclassed at the end

File: tests/trainer/test_tp.py (hunk 2 above)
Comment (eitanturok) [reply tests/trainer/test_tp.py:187]: just found it - `deep_compare`

Comment (eitanturok) [reply tests/trainer/test_tp.py]: I now use deep compare everywhere

File: tests/checkpoint/test_state_dict.py [hunk 4]
```diff
@@ -60,10 +60,10 @@ def test_get_model_state_dict_ignore(use_composer_model: bool):
         model = EvenSimplerMLP(num_features=8, device='cuda')
 
     model_state_dict = get_model_state_dict(model, sharded_state_dict=False, ignore_keys='module.2.weight')
-    assert set(model_state_dict.keys()) == {'module.0.weight'}
+    assert set(model_state_dict.keys()) == {'fc1.weight', 'fc2.weight', 'module.0.weight'}
```
Comment (mvpatel2000) [inline tests/checkpoint/test_state_dict.py:63]: why changes in this test?

File: tests/common/models.py [hunk 5]
```diff
@@ -122,10 +122,10 @@ class EvenSimplerMLP(torch.nn.Module):
 
     def __init__(self, num_features: int, device: str = 'cpu', num_out_features: int = 3):
         super().__init__()
-        fc1 = torch.nn.Linear(num_features, num_features, device=device, bias=False)
-        fc2 = torch.nn.Linear(num_features, num_out_features, device=device, bias=False)
+        self.fc1 = torch.nn.Linear(num_features, num_features, device=device, bias=False)
+        self.fc2 = torch.nn.Linear(num_features, num_out_features, device=device, bias=False)
```
Comment (mvpatel2000) [inline tests/common/models.py:126]: why attach to classes? this might be causing ur failure -- we intentionally removed it as it caused issues with FSDP

File: tests/trainer/test_tp.py [hunk 6]
```diff
@@ -1,27 +1,360 @@
 # Copyright 2022 MosaicML Composer authors
 # SPDX-License-Identifier: Apache-2.0
 
+from typing import Any, Optional, TypeVar, Union
+
+E = TypeVar('E', bound=BaseException)
```
Comment (mvpatel2000) [inline tests/trainer/test_tp.py:6]: why do we need this?

File: tests/trainer/test_tp.py [hunk 7]
```diff
@@ -1,27 +1,360 @@
...
+    replication: Optional[int] = None,
+):
+    fsdp_config = FSDPConfig(
+        state_dict_type='full',
+        sharding_strategy='SHARD_GRAD_OP',
+        mixed_precision='full',
+        use_orig_params=True,
+    )
+    parallelism_config = ParallelismConfig(fsdp=fsdp_config)
+
```
Comment (mvpatel2000) [inline tests/trainer/test_tp.py:116]: can fsdp + ddp + tp trainers all be one fn that has an arg with parallelism strat to reduce code copying?

**Review by mvpatel2000 — COMMENTED** [review]
some minor comments, mostly LGTM

File: tests/trainer/test_tp.py (hunk 7 above)
Comment (eitanturok) [reply tests/trainer/test_tp.py]: done!

File: tests/trainer/test_tp.py (hunk 6 above)
Comment (eitanturok) [reply tests/trainer/test_tp.py:7]: removed

File: tests/common/models.py (hunk 5 above)
Comment (eitanturok) [reply tests/common/models.py]: To resolve this, I created `TPSimpleComposerMLP` which is just like `SimpleComposerMLP` but saves the each layer. Then I stopped saving the layer in `SimpleComposerMLP`.

File: tests/checkpoint/test_state_dict.py (hunk 4 above)
Comment (eitanturok) [reply tests/checkpoint/test_state_dict.py]: I removed these changes.

File: tests/trainer/test_tp.py [hunk 8]
```diff
@@ -2,22 +2,333 @@
...
+    if parallelism_strategy == 'ddp':
+        trainer = get_base_trainer(
+            size=size,
+            batch_size=batch_size,
+            num_classes=num_classes,
+            num_features=num_features,
+            seed=seed,
+            device=device,
+            replication=replication,
+        )
```
Comment (mvpatel2000) [inline tests/trainer/test_tp.py:99]: instead of calling `get_trainer` 3 times, it should be one call and the if statement just creates parallelism config. For DDP, it can be None

Comment (eitanturok) [reply tests/trainer/test_tp.py]: Great idea! Done.

**Review by mvpatel2000 — APPROVED** [review]
